stan/math/rev/mat/functor/adj_jac_apply.hpp

```diff
@@ -48,6 +48,106 @@ template <class F, class Tuple>
 constexpr auto apply(const F& f, const Tuple& t) {
   return apply_impl(f, t, std::make_index_sequence<std::tuple_size<Tuple>{}>{});
 }
+
+/**
+ * build_y_adj takes the adjoint from the vari pointed to
+ * by y_vi_[0] and stores it in y_adj
+ *
+ * @tparam size dimensionality of M_
+ * @param y_vi_ pointer to pointer to vari
+ * @param M_ ignored in this specialization
+ * @param[out] y_adj reference to variable where adjoint is to be stored
+ */
+template <size_t size>
+void build_y_adj(vari** y_vi_, const std::array<int, size>& M_, double& y_adj) {
+  y_adj = y_vi_[0]->adj_;
+}
+
+/**
+ * build_y_adj takes the adjoints from the varis pointed to
+ * by y_vi_ and stores them in y_adj
+ *
+ * @tparam size dimensionality of M_
+ * @param y_vi_ pointer to pointers to varis
+ * @param M_ shape of y_adj
+ * @param[out] y_adj reference to std::vector where adjoints are to be stored
+ */
+template <size_t size>
+void build_y_adj(vari** y_vi_, const std::array<int, size>& M_,
+                 std::vector<double>& y_adj) {
+  y_adj.resize(M_[0]);
+  for (int m = 0; m < M_[0]; ++m)
+    y_adj[m] = y_vi_[m]->adj_;
+}
+
+/**
+ * build_y_adj takes the adjoints from the varis pointed to
+ * by y_vi_ and stores them in y_adj
+ *
+ * @tparam size dimensionality of M_
+ * @param y_vi_ pointer to pointers to varis
+ * @param M_ shape of y_adj
+ * @param[out] y_adj reference to Eigen::Matrix where adjoints are to be stored
+ */
+template <size_t size, int R, int C>
+void build_y_adj(vari** y_vi_, const std::array<int, size>& M_,
+                 Eigen::Matrix<double, R, C>& y_adj) {
+  y_adj.resize(M_[0], M_[1]);
+  for (int m = 0; m < M_[0] * M_[1]; ++m)
+    y_adj(m) = y_vi_[m]->adj_;
+}
+
+/**
+ * compute the dimensionality of the given template argument. By
```

Comment:
Capitalize sentences.

The bigger problem is that it's not sure what "compute the dimensionality" means here.  What's the dimensionality of a template argument?

Then this one should say that it's just a dummy base class with no default definition.  

I'm really not sure how to doc these base classes that are intended for specialization.